L1Trigger/L1THGCal/src/be_algorithms/HGCalMulticlusteringImpl.cc

```diff
@@ -70,15 +111,147 @@ void HGCalMulticlusteringImpl::clusterize( const edm::PtrVector<l1t::HGCalCluste
+            multiclustersTmp.at(i).set_sigmaRRMean(shape_.sigmaRRMean(multiclustersTmp.at(i)));
+            multiclustersTmp.at(i).set_eMax(shape_.eMax(multiclustersTmp.at(i)));
+
             multiclusters.push_back( 0, multiclustersTmp.at(i));  
         }
     }
     
 }
+void HGCalMulticlusteringImpl::clusterizeDBSCAN( const edm::PtrVector<l1t::HGCalCluster> & clustersPtrs, 
+                                                 l1t::HGCalMulticlusterBxCollection & multiclusters,
+                                                 const HGCalTriggerGeometryBase & triggerGeometry)
+{
+
+    std::vector<l1t::HGCalMulticluster> multiclustersTmp;
+    l1t::HGCalMulticluster mcluTmp;
+    std::vector<bool> visited(clustersPtrs.size(),false);
+    std::vector<bool> merged (clustersPtrs.size(),false);
+    std::vector<std::pair<unsigned int,double>>  rankedList;
+    rankedList.reserve(clustersPtrs.size());
+    std::vector<std::vector<unsigned int>> neighborList;
+    neighborList.reserve(clustersPtrs.size());
+
+    int iclu = 0, imclu = 0, neighNo;
+    double dist = 0.;
+
+    for(edm::PtrVector<l1t::HGCalCluster>::const_iterator clu = clustersPtrs.begin(); clu != clustersPtrs.end(); ++clu, ++iclu){
+        dist = (*clu)->centreProj().mag()*HGCalDetId((*clu)->detId()).zside();
+        rankedList.push_back(std::make_pair(iclu,dist));
+    }  
+    iclu = 0;
+    std::sort(rankedList.begin(), rankedList.end(), [](auto &left, auto &right) {
+            return left.second < right.second;
+            });
+
+    for(auto cluRanked: rankedList){
+        std::vector<unsigned int> neighbors;      
+
+        if(!visited.at(iclu)){
+            visited.at(iclu) = true;
+            findNeighbor(rankedList, iclu, clustersPtrs, neighbors);
+            neighborList.push_back(std::move(neighbors));
+
+            if(neighborList.at(iclu).size() >= minNDbscan_) {
+                multiclustersTmp.emplace_back( clustersPtrs[cluRanked.first] );
+                merged.at(iclu) = true;
+                /* dynamic range loop: range-based loop syntax cannot be employed */
+                for(unsigned int neighInd = 0; neighInd < neighborList.at(iclu).size(); neighInd++){
+                    neighNo = neighborList.at(iclu).at(neighInd);
+
+                    if(!visited.at(neighNo)){
+                        visited.at(neighNo) = true;
+                        std::vector<unsigned int> secNeighbors;
+                        findNeighbor(rankedList, neighNo,clustersPtrs, secNeighbors);
+                        multiclustersTmp.at(imclu).addConstituent( clustersPtrs[rankedList.at(neighNo).first]);
+                        merged.at(neighNo) = true;
+
+                        if(secNeighbors.size() >= minNDbscan_){
+                            neighborList.at(iclu).insert(neighborList.at(iclu).end(), secNeighbors.begin(), secNeighbors.end());
+                        }
+
+                    } else if(!merged.at(neighNo) ){
```

Comment:
There is one case where it may have an impact, if the cluster has been merged but not visited. But I'm not sure if this case is possible. 
@osahin could you check if the change proposed above is strictly equivalent to what is currently implemented? Thanks!

Reply:
@osahin ping

Reply:
Hi @kpedro88, sorry I missed this mention due to the end of year break. I agree that the suggested change would shorten the code, yet now thinking back a bit, I intentionally left it this way to improve the readability (ie to put an explicit case for the clusters that are already visited by other multicluster seeds but not merged). Please correct me if I am wrong, but the suggested change would not save cycles, so keeping it makes the code a bit easier to digest. I would be more than happy to further discuss it.

Reply:
My concern whenever code is duplicated is that some day in the future, another developer will update one case but not the other, leading to inconsistent behavior. To me, the code would still be quite comprehensible if `!visited` was explicitly a subcase of `!merged`.